### ISSUE TITLE
[BUG] Trying to commit an action with an app validation that depends on a non-cached dependency fails (which happens quite often if the node trying to commit has "empty" gossip arc clamping)

### ISSUE THREAD

**Issue by guillemcordoba** [holochain/holochain#3901]
**Describe the bug**

When a node with "empty" gossip arc clamping creates a link and then tries to delete it, it gets this validation failure:

```
called `Result::unwrap()` on an `Err` value: CellError(WorkflowError(SourceChainError(InvalidCommit("Awaiting deps [AnyDhtHash(uhCkkpCE1YL36QAoCYgSBYKf9ij3huj_GHTUG2a8NCZ8xelFS1kCT)] but this is not allowed when committing entries to the source chain")))),
```

**Expected behavior**

Nodes with "empty" gossip arc clamping should be able to call `delete_link` on the links they created.

**System information:**
 - OS: Ubuntu 22.04
 - Holochain and HDK Version (run `hn-introspect` as well as `holochain --build-info`
 from the nix shell and paste output)

`hn-introspect`:

holochain (holochain 0.3.1-rc.0): 9f293072c41bd053e9cd88f746a456ee1dee1990
lair-keystore (lair_keystore 0.4.4): 826be915efc839d1d1b8a2156b158999b8de8d5b
hc-launch (holochain_cli_launch 0.300.0-dev.1 (holochain 0.3.0-beta-dev.48)): b96d5aa790bf0da0a8f0c44741276f7f9c4b6b41
hc-scaffold (holochain_scaffolding_cli 0.3000.0-dev.9): 5c5e19cad570046291c7526da780d1c94d13918f

`holochain --build-info`
{"git_info":null,"cargo_pkg_version":"0.3.1-rc.0","hdk_version_req":"0.3.1-rc.0","hdi_version_req":"0.4.1-rc.0","lair_keystore_version_req":"0.4.4","timestamp":"2024-05-20T18:40:34.733972908Z","hostname":"localhost","host":"x86_64-unknown-linux-gnu","target":"x86_64-unknown-linux-gnu","rustc_version":"rustc 1.77.2 (25ef9e3d8 2024-04-09)","rustflags":"","profile":"release"}

**Steps to reproduce**

Here you have a reproduction of the bug, which is a scaffolded app that has the ability to call delete link, with a successful test that calls delete_link with default gossip arc clamp, and a failing test that has "empty" gossip arc clamping.

https://github.com/guillemcordoba/empty-arc-clamping-bug-reproduction

**Additional context**

This is blocking holochain from fully working in mobile.

**Comment by abe-njama** [comment]
@maackle could this be a regression that was introduced through the fix that [refactors Kitsune to use quantized arcs](https://github.com/holochain/holochain/pull/3497)?

**Comment by ThetaSinner** [comment]
This error happens within the core workflows when app validation requests a hash that isn't found locally - there is never an option to go to the network during the app validation self-check.

I agree that deleting a link you created should work but this doesn't go as far as Kitsune, so does this definitely work when gossip is configured another way and nothing else is different?

Is the hash that app validation is claiming is missing definitely the same as the create-link hash? I'm wondering if something else has ended up getting requested in app validation that is causing trouble with needing to fetch it

**Comment by guillemcordoba** [comment]
You can see in the test folder that I created two identical tests except the gossip arc clamping configuration, which give different results:
https://github.com/guillemcordoba/empty-arc-clamping-bug-reproduction/tree/main/dnas/forum/zomes/coordinator/posts/tests

**Comment by neonphog** [comment]
@guillemcordoba - I've tried to reproduce your reproduction as faithfully as possible: https://github.com/holochain/holochain/pull/3967

It is not showing the bug. Did I miss some critical step in your reproduction? Or, is it possible this bug got fixed by some other change in the mean time?

Thanks for checking!

**Comment by guillemcordoba** [comment]
Hey @neonphog ! Okey so I think I understand better the issue now. Yes it's still present in the 0.4 branches of holochain, here is a PR that adds app validation to the zome that you created to reproduce the issue: https://github.com/holochain/holochain/pull/4000.

So, I think this is the issue: https://github.com/holochain/holochain/pull/3433 made it so that validation does not fetch dependencies immediately, only logs that it needs them and retries later. This is fine for the app validation workflow, but I don't think it is for the inline validation that happens at the end of the call zome workflow. There, we are counting on the node actually having in its dht store or _in cache_ all the necessary dependencies for the commit to succeed. This happens very often in small networks, where there is no sharding. But, in big networks with sharding, or in networks where some node has "empty" gossip arc clamping, this is not guaranteed to be the case: some node with empty gossip arc clamping could fetch some data, wait some time until the cache is cleared, and then try to commit the dependant action. I don't think we can rely on the cache for such an important mechanism (the UX for an app completely breaks down, as you can't delete a post for example, even if you are seeing it in your UI).

I actually was able to reproduce the same bug without setting the second node's gossip arc clamping to zero: just try to commit some dependant action just after the node that created the action that is depended upon can fetch it.

Ping @jost-s.

**Comment by jost-s** [comment]
Hi @guillemcordoba. We're thinking of solving this by allowing inline validation to make network calls.

What confuses me is that you see this happen often. It only applies to dependencies that haven't been fetched during inline sys validation, which can make network calls. What inline sys validation wouldn't fetch are a link base when deleting a link. And to delete a link, an agent must know its hash. So in which scenario is this a frequent event, that an agent is deleting a link it doesn't have the create hash of?

**Comment by guillemcordoba** [comment]
Yeah it _must_ have the create hash of it, but it _may not_ have the actual record for that hash cached. That's the difference and what I'm referring to here. I don't know the internals of the caching mechanisms well enough, but it seems to me like it can very well be the case that you have access to the hash and try to commit something with it, without having the record in cache because you haven't needed to fetch it lately.

**Comment by jost-s** [comment]
Okay, I see. Even though it's called "cache", it's never invalidated/emptied at the moment. So whenever you fetch something and it lands in your cache, it'll stay.

I suppose this has happened while using an app on a phone?

**Comment by guillemcordoba** [comment]
Yeap, leading to terrible UX. It happened in my phone I think because I didn't fetch the create record, but just signaled it to my UI. And then I tried to delete it from the UI and couldn't as a result of this error. 

I think fixing this issue should fix what I saw in my phone, but I'll have to try again after it's fixed to be certain. Nevertheless it's an actual issue that should be fixed IMO.

**Comment by jost-s** [comment]
It seems problematic that the UI displays data it does not have in a storage. If that were the case, you'd restart the UI and wouldn't have the data anymore.

**Comment by jost-s** [comment]
@guillemcordoba Would it not be a problem in an app that any data incoming through a signal is not persisted and lost after restarting the app?

**Comment by guillemcordoba** [comment]
Well maybe... But it depends on the app. There could be real-time games which don't really care about that, or the entry could be stored on the frontend side via local storage, or sent to some other node for persistence... Even storing the hash in a private entry and then only querying it and trying to do some action with it will be enough to trigger this issue.

I just tried it again with my phone and I think I made a mistake when describing the problem I was seeing with it. I actually found another bug! This is the actual bug that I found while testing things with my phone: https://github.com/holochain/holochain/issues/4012

I still think this bug (Trying to commit an action with an app validation that depends on a non-cached dependency fails) exists and it's important to address it. So now it's actually two issues instead of one.
